[DOC,BUG] Add note about FWHM to Morlet, fix norm

britta-wstnr opened the pull request:
This is a draft of a note that could be added to `tfr_morlet` on how to derive the FWHM for the wavelet that has been used.

Draft status because the equation reported in [Cohen 2019](https://www.sciencedirect.com/science/article/pii/S1053811919304409) is not delivering an accurate result on our wavelets, only an approximation.

Reasons for this are related to: https://mne.discourse.group/t/mne-pythons-implementation-of-morlet/5892

CC @drammock @larsoner

britta-wstnr commented:
@larsoner 
FWIW, here is a [gist](https://gist.github.com/britta-wstnr/c33a4e2e1d0854f56ad9b5eb3a1246f6) I used to play around with the width estimation. Maybe it can save you some time.

larsoner commented:
Pushed a commit, to me MNE and `scipy.signal.morlet2` appear equivalent up to a scale factor of `sqrt(2)`

larsoner commented:
https://output.circle-artifacts.com/output/job/4661b749-1f72-4fb7-bb2a-c5c95adf22d1/artifacts/0/dev/generated/mne.time_frequency.morlet.html

![](https://circleci-tasks-prod.s3.us-east-1.amazonaws.com/forks/storage/artifacts/c889073a-e9a4-4303-a8fa-a8929abb369f/86345521/4661b749-1f72-4fb7-bb2a-c5c95adf22d1/0/dev/generated/mne-time_frequency-morlet-1.hires.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQVFQINEODUN2PMJ6%2F20221205%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221205T172157Z&X-Amz-Expires=60&X-Amz-Security-Token=IQoJb3JpZ2luX2VjENr%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJGMEQCIGWDnS5bwawACNgtsdPM2xfEaBBcAN4jd9qob%2BSSs1p5AiAOkGRp2cqNOOfu%2FJkQmEkECqPf8NvB%2FbVI4dCwwCD74iq0Agjy%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAMaDDA0NTQ2NjgwNjU1NiIMspIPAxUaikUplX%2FwKogCfQWgF4aSxSMY8vCUZoZkDDbOlicWwE0Zq13ggq4niYA32tCyfNrm%2FLIXcaqYiLq1bPs%2Bzm7GyTFsvEA1wDKrQt%2F4pgL1r1SDlef9QK0rlgQHNtemZ4FUN1Np2pwN8q65Wb7r%2FH%2FjubudnU9hrQ9Bw%2FzL4P1F%2F3il8QgJ56iOaA%2F5TkG%2Br%2F0FBm77nLj2jdusyhshEtPMO8OyaS8oeMtRem0HdErfmzqzFn8XMFgSjuE76TPCd0WB0vHu4h7JQ6KU85a5KHytexrjd7q2XjsUGiNTjb107%2BemU7GWT24x7TGrhx9tCpzpDr8t%2BO9pA3qY0L5DgTCAxPypIJmBVHZ9HNWi8dDUujG9MNrLuJwGOp4BpGzK4K01%2FOZQfanAQZdyuaNiVyx1uohs7ipYmkGV1R7D7wuA9Xjs8%2FN%2BJ8satC71QZLydnwhLhxoNnRen4EiALPmpXAFD09Ie4T1ggRT3vQJJ6jqCnPltK65tRWriNCaP4vzqtdNUjNfGkBkqt4OCRoxfFS56IDaNwPHKg%2BqFOpsOdUHgbHgHAYWPiLFkbGXDf2ikZwldRx8otpHzRU%3D&X-Amz-SignedHeaders=host&x-id=GetObject&X-Amz-Signature=02eb08887aec46e06ca8a89f338f403022e7f46dd6d17d4e165d9873f1186241)

larsoner commented:
... to me the one possible bugfix would be to change this:
```
W /= np.sqrt(0.5) * np.linalg.norm(W.ravel())
```
I don't understand why there is a multiplication by `np.sqrt(2)` (accomplished by division by `sqrt(0.5)`). It seems like the atom should be unit power like in SciPy. @agramfort you added this normalization 12 years ago (!), let me know if you agree we should remove the `np.sqrt(2)` multiplication factor so we can obtain `np.linalg.norm(W) == 1` instead of `np.linalg.norm(W) == np.sqrt(2)` (as it is in main).

larsoner commented:
Can you check if the norm of the complex wavelet in FieldTrip has norm 1 or norm `sqrt(2)`? If FieldTrip and SciPy both have norm 1 then I think we should fix our code...

larsoner commented:
I'm starting to convince myself that we're wrong:
```
>>> from scipy.signal import cwt
>>> from mne.time_frequency import tfr_array_morlet, morlet
>>> import numpy as np
>>> x = np.zeros(10001)
>>> x[5000] = 1  # delta
>>> np.linalg.norm(x)
1.0
>>> m = tfr_array_morlet(x[np.newaxis, np.newaxis], freqs=[10.], sfreq=1000., n_cycles=7)[0][0][0]
>>> m.shape
(10001,)
>>> np.linalg.norm(m)  # does not preserve norm, goes to sqrt(2)
1.4142135623730951
>>> M = morlet(sfreq=1000., freqs=10., n_cycles=7).size
>>> s = 7 * 1000. / (2 * 10 * np.pi)
>>> sp = cwt(x, morlet2, widths=[s], w=7)[0]
>>> np.linalg.norm(sp)  # does preserve norm
0.9999999999992621
```

britta-wstnr commented:
@larsoner I don't have access to MATLAB right now but can check tomorrow

larsoner commented:
@britta-wstnr thinking about it more, I'm not sure it matters if `norm(wavelet) == 1` for FieldTrip -- I think we should be consistent with SciPy's `cwt` here. (Also the norm being 1. was my initial expectation before looking at SciPy.)

So @britta-wstnr @drammock if you're both happy I think we can merge this as a bugfix + doc update

drammock commented on mne/time_frequency/tests/test_tfr.py:
```diff
@@ -38,13 +39,35 @@ def test_tfr_ctf():
+    # Check FWHM
+    fwhm_formula = n_cycles * np.sqrt(2 * np.log(2)) / (np.pi * freq)
+    half_max = np.abs(W).max() / 2.
+    fwhm_empircal = (np.abs(W) >= half_max).sum() / sfreq
+    # Could be off by a few samples
+    assert_allclose(fwhm_formula, fwhm_empircal, atol=3 / sfreq)
```
```suggestion
    fwhm_empirical = (np.abs(W) >= half_max).sum() / sfreq
    # Could be off by a few samples
    assert_allclose(fwhm_formula, fwhm_empirical, atol=3 / sfreq)
```

drammock commented on mne/utils/docs.py:
```diff
@@ -1417,6 +1417,19 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
+wavelet can be reported.
+
+The FWHM of the wavelet at a specific frequency is defined as:
+:math:`FWHM = \frac{n\_cycles \sqrt{2 \ln{2}}}{\pi freq}` (cf. eq. 4 in
+:footcite:`Cohen2019`).
+"""
```
```suggestion
:math:`\mathrm{FWHM} = \frac{\mathtt{n\_cycles} \times \sqrt{2 \ln{2}}}{\pi \times \mathtt{freq}}`
(cf. eq. 4 in :footcite:`Cohen2019`).
"""  # noqa E501
```

drammock reviewed (approved):
LGTM, just a spelling fix and some equation formatting.

What happened to the idea of the FWHM computation becoming a helper function?  Simply forgotten, or YAGNI? (it shows up in both the example and in the tests, and since we're recommending it to users as a better thing to report, I suspect it would get used).

britta-wstnr commented:
So I did some FieldTrip-code-digging after all and also partly re-read Tallon-Baudry et al. (1997).

1. FieldTrip has a parameter `gwidth`, which relates to our 5.0 - the one that started off the discussion over at Discourse. Its default equates to 3. It determines the width of the Gaussian, but not directly of the wavelet, as this gets zero-padded.
2. FieldTrip does not scale the wavelet such that norm 1.
3. FieldTrip scales by A, with A being:
```matlab
  A = 1 / sqrt(sigma_t * sqrt(pi));
  ```
  which is in line with Tallon-Baudry et al - and what we are doing on `main`, I believe?
  
  Relevant part from the paper:
  
<img width="426" alt="image" src="https://user-images.githubusercontent.com/20129119/206008771-ed8f4b2d-d0cc-4f6a-b2fd-a397e44f08d6.png">
(Tallon-Baudry et al. 1997, Journal of Neuroscience, p. 724)

larsoner commented:
> which is in line with Tallon-Baudry et al - and what we are doing on main, I believe?

To me it looks like we have 1) written that formula in a comment, and 2) claimed that whatever we do is equivalent to using it. So yes I think we *intend* to use something equivalent to it at least 🙂 But I don't see that term show up in our actual code in `main`:

https://github.com/mne-tools/mne-python/blob/3812114eeea9a9ffc8449fb23b21353e878e6f25/mne/time_frequency/tfr.py#L90-L100

So to work it out a bit:

1. Empirically, the `norm` of the wavelet on `main` is currently always `sqrt(2)`, not `1.` (which is what it is on this PR) -- so if by "energy" they mean "signal norm" (or taking the norm to any power) here, removing the `sqrt(2)` multiplication seems like the right move.

2. Analytically, looking at SciPy's implementation, they use:

    https://github.com/scipy/scipy/blob/684a519a327f76cdc4d9eee2a88d99437b90395f/scipy/signal/_wavelets.py#L400-L401

    so for them, `A=np.sqrt(1/s) * np.pi**(-0.25)` effectively, I think, which is the same as FieldTrip's `A=1/sqrt(s*sqrt(pi))`. So by changing to a norm of `1.` rather than `sqrt(2)` we're consistent with SciPy, which seems to be consistent with FieldTrip if I'm interpreting what you wrote correctly.

So @britta-wstnr when you said:

> FieldTrip does not scale the wavelet such that norm 1.

Did you mean that the resulting wavelet did not have norm (numerically) equal to 1 when you checked, or that by looking at the source code there is no division by `norm(wavelet)` but rather this analytic term? If the latter, we should check -- I suspect that the analytical normalization in Tallon-Beaudry and FieldTrip (and I think SciPy) effectively ensures a unit norm in the end, aside from low-`n_cycles` / intergation issues. (And I *think* our empirical normalization is in practice perhaps better because it avoids issues with numerical approximation / low-n-cycles, but maybe not...)

drammock commented:
on main, the same normalization (`W /= sqrt(2) * np.linalg.norm(W.ravel())`) is also done in the `_make_dpss()` function.  So although removing the `sqrt(2)` here does seem to make sense, I think we should be really thorough here to make sure we're not creating more problems for ourselves (matching SciPy is probably good, but at the expense of our morlet and dpss having different properties?).

drammock commented:
Here are some notes I made early on to help myself understand all this (which I'm still not sure I do).  Copy-paste to a wide window to get the terms on each line to align nicely:

```
                           gaussian window              sinusoidal part                           energy normalization constant
                          ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓    ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓    ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓
MNE:                      exp(-0.5*(t / σ_t)^2) *  exp(1j * f *  t * 2*pi)              * sqrt(2) / norm(W.ravel())
MATLAB (s=√(bandwidth))   exp(     ( x / s )^2) *  exp(1j * f *  t * 2*pi)              *             sqrt(1 / (pi * s^2))
scipy morlet2:            exp(-0.5*( x / s )^2) *  exp(1j * w * x/s)                    * pi^(-1/4) * sqrt(1 / s)
scipy morlet (standard):  exp(-0.5*    x    ^2) *  exp(1j * w *  x )                    * pi^(-1/4)
scipy morlet (complete):  exp(-0.5*    x    ^2) * (exp(1j * w *  x ) - exp(-0.5*(w^2))) * pi^(-1/4)
wikipedia:                exp(-0.5*    t    ^2) * (exp(1j * σ *  t ) - exp(-0.5*(σ^2))) * pi^(-1/4) * sqrt(1 / (1 + exp(-σ^2) - 2*exp(-3/4*σ^2)))
                                                                       ↑↑↑↑↑ κ_σ ↑↑↑↑↑
notes:
- the sqrt(...) part of the energy normalization constant is very nearly 1 as long as omega0 is big enough (≥5)
- the κ_σ part is required for "admissibility" (for the wavelet to work as a mother wavelet in the CWT).
  Like the sqrt normalization constant term, κ_σ also gets very small as omega0 increases (if omega0=5, κ_σ is ~4e-6)
  and so κ_σ is often omitted as well.
```

britta-wstnr commented:
@larsoner, actually, FieldTrip does normalize - just way later, my bad. They normalize by `sqrt(2 ./ fsample)` - in the frequency domain, it seems. I can look more into that tomorrow ... 🌚

larsoner commented:
> then norm(wavelet_fieldtrip) = norm(wavelet_mne_main) = sqrt(2)

Okay I think you might be talking about the output of `ft_specest_wavelet` and comparing to `tfr_array_morlet`, right?

Earlier in the code we do see a unit norm from the analytical normalization of the `morlet` wavelet itself
```
for ifreqoi = 1:nfreqoi
  dt = 1/fsample;
  sf = freqoi(ifreqoi) / width(ifreqoi);
  st = 1/(2*pi*sf);
  toi2 = -gwidth*st:dt:gwidth*st;
  A = 1/sqrt(st*sqrt(pi));
  tap = (A*exp(-toi2.^2/(2*st^2)))';
  disp(norm(tap))
  error('bad')
```
I get in Octave:
```
octave:4> dat = zeros(10000,1);
octave:5> dat(5000,1)=1;
octave:6> norm(dat)
ans = 1
octave:8> ft_specest_wavelet(dat, 1:10000)
1.0000
error: bad
```
so `tap` -- which I think contains the CWT? -- has norm 1. for me. Which is good, because I think it means this PR's MNE, FieldTrip, and SciPy all agree on how a `morlet` wavelet itself is normalized. But...

> FieldTrip does normalize - just way later, my bad. They normalize by sqrt(2 ./ fsample)

It looks like this either doesn't matter, or falls out, because changing `fsample` does not appear to modify it:
```
> dat = zeros(1,10000);
> dat(1, 5000) = 1;
> [spectrum, freqoi, timeoi] = ft_specest_wavelet(dat, 1:10000, 'freqoi', [0.01]);
> norm(spectrum(:, 3000:7000))
ans = 1.4142
> [spectrum, freqoi, timeoi] = ft_specest_wavelet(dat, (1:10000) / 500, 'freqoi', [0.01 * 500]);  % 500 Hz equiv
frequency 1 (5.00 Hz)
> norm(spectrum(:, 3000:7000))
ans = 1.4142
```
So MNE+FieldTrip agree in the `cwt` output itself on `main` with SciPy being smaller by a factor of sqrt(2), whereas MNE+SciPy agree on this PR with FieldTrip being bigger by a factor of `sqrt(2)`. Because SciPy preserves the norm of the wavelet when convolved with an impluse:
```
>>> np.linalg.norm(cwt(x, morlet2, [10], w=5))
0.9999999999992631
```
To me it's a bug in FieldTrip that the convolution of the wavelet with an impulse does not yield the wavelet with norm 1 anymore...

larsoner commented:
> on main, the same normalization (W /= sqrt(2) * np.linalg.norm(W.ravel())) is also done in the _make_dpss() function.

Argh yes, even though the DPSS window function does not have it natively we seem to add it there:
```
>>> np.linalg.norm(mne.time_frequency.multitaper.dpss_windows(10000, 10, 1)[0])
0.9999999999999999
>>> np.linalg.norm(mne.time_frequency.tfr._make_dpss(1000., [10])[0][0])
1.4142135623730951
```
So maybe it is a factor we should put in when computing the CWT, but not in `morlet` itself. This would at least be the least invasive, most consistent change. Then we could punt on the `sqrt(2)` stuff until another PR when someone looks deeper into the math :)

britta-wstnr commented:
I took another dive into the math (Tallon-Baudry et al., 1997) and FieldTrip as well as MNE-Python.
I am now convinced that we are right here.

**TLDR:** `main` seems to be golden if we go by Tallon-Baudry 1997 and FieldTrip.

In more detail:

## Taper, envelope, and norm
The taper or wavelet norm in FieldTrip is not 1 when I debug it. It is first dependent on the sampling frequency, and when multiplied with normalizing in line https://github.com/fieldtrip/fieldtrip/blame/master/specest/ft_specest_wavelet.m#L228, the norm of the wavelet is `sqrt(2)`. The same is true for MNE-Python.

## Normalization of the taper
FieldTrip multiplies the taper with `A = 1 / np.sqrt(sigma_t * np.sqrt(np.pi))`. This is following the wavelet definition in Tallon-Baudry et al. MNE-Python does not do this, but line https://github.com/mne-tools/mne-python/blob/main/mne/time_frequency/tfr.py#L100 achieves the same thing: the normalization here becomes 1 if we add A - i.e., if we add
 ```python
a = 1 / np.sqrt(sigma_t * np.sqrt(np.pi));
gaussian_enveloppe = a * np.exp(-t ** 2 / (2.0 * sigma_t ** 2))
```
all tests still pass.

## Length of the oscillation
The 5 does not matter much (I tried to check out some extreme cases). It adds more cycles to the frequency than initially necessary, but this gets "corrected" by the taper (all those extra cycles get forced to 0). 
Setting this to 4 instead of 5 makes one tests fail. -- FieldTrip sets it to 3 by default, but it can be adapted by the user.

## Comparison MNE (stable release) with FieldTrip
Below the output of using `ft_specest_wavelet()` and `tfr_array_morlet()` on a 500 Hz sampling rate, 2 second time series with one impulse, reconstructing 10 Hz at 7 cycles:

<img width="563" alt="image" src="https://user-images.githubusercontent.com/20129119/206533328-2e42307a-e8e6-464b-9257-d2a98b8b5655.png">
<img width="390" alt="image" src="https://user-images.githubusercontent.com/20129119/206533391-d263983d-00e1-4f92-9362-4184723e27b8.png">

britta-wstnr commented:
@larsoner that would be line https://github.com/fieldtrip/fieldtrip/blame/master/specest/ft_specest_wavelet.m#L164
That code apparently hasn't changed in 12 years ... but depends on how old your FieldTrip is I guess 😆 

I meant on `main`- at least if we go after Tallon-Baudry I think. I have not checked this PR yet.

larsoner commented:
> @larsoner that would be line https://github.com/fieldtrip/fieldtrip/blame/master/specest/ft_specest_wavelet.m#L164

Yes that's the spot -- did you check there? For me that norm is `1`, so it should be for you, too. If I'm right that that's the wavelet itself, then I think the approach on this PR -- having the wavelet be norm 1 but the `cwt` output a norm of sqrt(2) -- seems consistent with FieldTrip and tallon-beaudry

larsoner commented:
Ahh makes sense. So at the end of the day do think `morlet` should have an output of norm `sqrt(2)` rather than 1.?

britta-wstnr commented:
I think - that it all comes down to the following sentence from Tallon-Baudry and whether we want this to be true:
> Wavelets are normalized so that their total energy is 1, the normalization factor A being equal to: `(sigma_t * np.sqrt(pi))**0.5`

This is true for the real part of the wavelet - and as far as I can see - this is the difference to other wavelet formulations, right?

I would argue we want to follow Tallon-Baudry, as this is the heavily quoted paper for wavelets in neuroscience ...

larsoner commented:
> To me the "wavelet" is the complex thing, not just the real or imaginary part

Ahh yes it makes sense now that this is the difference. It's a question of whether the real part and imag parts by themselves should have norm 1, or the complex `real+j*imag` should have norm 1. I'll revert the `sqrt(2)` changes and adjust the tests back, that way we can stay consistent as you suggest :+1:

britta-wstnr commented:
Note on FWHM: Digging deeper into it, it seems like Cohen 2019 (reference above) does _not_ normalize the wavelet by _A_ in his paper. Thus, the FWHM computed based on his formula does not match the actual FWHM of our taper ...
Cohen cites Tallon-Baudry (1999; DOI: https://doi.org/10.1016/S1364-6613(99)01299-1) in his paper, and a quick look also reveals normalization factor _A_ there ... 
So if someone is up for double checking the math on that (Cohen eq. 3 vs Tallon-Baudry et al. 1997 or 1999) - on this side of the globe it's time to log off now ... 🌚

larsoner commented:
For now I'm just going to revert to `sqrt(2)` norm and call it good enough :) That way the `fwhm` stuff is in and working and we can look into the math more when someone is bored without holding up the upcoming 1.3 release

britta-wstnr commented:
But it is only approximately right ... no? FWHM, I  mean.

larsoner commented on mne/time_frequency/tests/test_tfr.py:
```diff
@@ -38,13 +39,35 @@ def test_tfr_ctf():
+    half_max = np.abs(W).max() / 2.
+    fwhm_empirical = (np.abs(W) >= half_max).sum() / sfreq
+    # Could be off by a few samples
+    assert_allclose(fwhm_formula, fwhm_empirical, atol=3 / sfreq)
```
@britta-wstnr this asserts that `fwhm` is correct to numerical precision (given the rough "half max" estimation mode used here at least, i.e., up to 3 samples off at the given `sfreq`)

larsoner reviewed (commented):
... and the updated example shows it's correct as well (from my local doc build that I'm about to push -- this will end up in the `morlet` doc):

![image](https://user-images.githubusercontent.com/2365790/206562181-3ad7e648-fbf1-4794-aeab-5888f2357a68.png)


Perhaps you and @drammock were not operating on the `abs(wavelet)` but rather just the real or imaginary part when you investigated? That would make it much harder to say what the `fwhm` actually is/means.

larsoner commented:
> Note on FWHM: Digging deeper into it, it seems like Cohen 2019 (reference above) does not normalize the wavelet by A in his paper. Thus, the FWHM computed based on his formula does not match the actual FWHM of our taper ...

You're talking about height/vertical normalization, right? Normalization/amplitude of the wavelet in that way shouldn't matter for FWHM, because it's a normalized measure itself -- if you rescale the thing you're computing the FWHM of (by anything other than 0), the max changes, so the half-max changes... and the FWHM is thus unchanged.

drammock commented:
I can do a thorough review of this tomorrow (Friday).

britta-wstnr commented:
> You're talking about height/vertical normalization, right? Normalization/amplitude of the wavelet in that way shouldn't matter for FWHM, because it's a normalized measure itself -- if you rescale the thing you're computing the FWHM of (by anything other than 0), the max changes, so the half-max changes... and the FWHM is thus unchanged.

Ah, yes of course. I double checked and indeed, both tapers have the same FWHM - I was fooled by their dramatically different look!

britta-wstnr commented:
I added the reference to the Tallon-Baudry paper and also updated the comment in the `morlet` function to hopefully make it clearer for future maintenance.

larsoner reviewed (approved):
@drammock feel free to mark as merge-when-green if you're happy!

drammock commented:
I pushed some comment-only changes to `morlet()`, added a reference to `fwhm()` docstring, and ignored another `Widget` deprecation. Good to go by me; enabling auto-merge. Thx @britta-wstnr and @larsoner !!
